pkg/solana/txm/pendingtx.go

```diff
@@ -3,78 +3,122 @@ package txm
+	Expired(sig solana.Signature, confirmationTimeout time.Duration) bool
+	// OnProcessed marks transactions as Processed
+	OnProcessed(sig solana.Signature) (string, error)
+	// OnConfirmed marks transaction as Confirmed and moves it from broadcast map to confirmed map
+	OnConfirmed(sig solana.Signature) (string, error)
+	// OnFinalized marks transaction as Finalized, moves it from confirmed map to finalized map, removes signatures from signature map to stop confirmation checks
+	OnFinalized(sig solana.Signature, retentionTimeout time.Duration) (string, error)
+	// OnError marks transaction as errored, matches err type using enum, moves it from confirmed map to finalized map, removes signatures from signature map to stop confirmation checks
+	OnError(sig solana.Signature, retentionTimeout time.Duration, errType int) string
+	// GetTxState returns the transaction state for the provided ID if it exists
+	GetTxState(id string) (TxState, error)
+	// TrimFinalizedErroredTxs removes transactions that have reaches their retention time
+	TrimFinalizedErroredTxs()
+}
+
+type pendingTx struct {
+	tx          solana.Transaction
+	cfg         TxConfig
+	signatures  []solana.Signature
+	id          string
+	createTs    time.Time
+	retentionTs time.Time
+	state       TxState
 }
 
 var _ PendingTxContext = &pendingTxContext{}
 
 type pendingTxContext struct {
-	cancelBy  map[uuid.UUID]context.CancelFunc
-	timestamp map[uuid.UUID]time.Time
-	sigToID   map[solana.Signature]uuid.UUID
-	idToSigs  map[uuid.UUID][]solana.Signature
-	lock      sync.RWMutex
+	cancelBy map[string]context.CancelFunc
+	sigToID  map[solana.Signature]string
+
+	broadcastedTxs      map[string]pendingTx // transactions that require retry and bumping i.e broadcasted, processed
+	confirmedTxs        map[string]pendingTx // transactions that require monitoring for re-org
+	finalizedErroredTxs map[string]pendingTx // finalized and errored transactions held onto for status
+
+	lock sync.RWMutex
 }
 
 func newPendingTxContext() *pendingTxContext {
 	return &pendingTxContext{
-		cancelBy:  map[uuid.UUID]context.CancelFunc{},
-		timestamp: map[uuid.UUID]time.Time{},
-		sigToID:   map[solana.Signature]uuid.UUID{},
-		idToSigs:  map[uuid.UUID][]solana.Signature{},
+		cancelBy: map[string]context.CancelFunc{},
+		sigToID:  map[solana.Signature]string{},
+
+		broadcastedTxs:      map[string]pendingTx{},
+		confirmedTxs:        map[string]pendingTx{},
+		finalizedErroredTxs: map[string]pendingTx{},
 	}
 }
 
-func (c *pendingTxContext) New(sig solana.Signature, cancel context.CancelFunc) (uuid.UUID, error) {
-	// validate signature does not exist
+func (c *pendingTxContext) New(tx pendingTx, sig solana.Signature, cancel context.CancelFunc) error {
 	c.lock.RLock()
+	// validate signature does not exist
 	if _, exists := c.sigToID[sig]; exists {
 		c.lock.RUnlock()
-		return uuid.UUID{}, errors.New("signature already exists")
+		return errors.New("signature already exists")
```

Comment:
nit: could define a constant for errors that are used multiple times across the file:
- signature already exists
- tx id already exists in broadcasted map
- id does not exist in broadcasted map
- id does not exist in broadcasted or confirmed map

Reply:
Good idea, I added some more predefined errors in the latest commit. I ended up consolidating the `id does not exist...` ones into a single error. Think the stack trace with the error can be used to determine which map(s) it was missing from to keep things simpler.